File: api/vedic.py
"""
Vedic Astrology Calculator
Uses Swiss Ephemeris (pyswisseph) for sidereal planetary positions.
Rule-based — no AI involved in calculation.
"""

import swisseph as swe
import math
import os
from datetime import datetime, timezone, timedelta

SIGNS = ["Aries", "Taurus", "Gemini", "Cancer", "Leo", "Virgo",
         "Libra", "Scorpio", "Sagittarius", "Capricorn", "Aquarius", "Pisces"]

NAKSHATRAS = [
    "Ashwini", "Bharani", "Krittika", "Rohini", "Mrigashira", "Ardra",
    "Punarvasu", "Pushya", "Ashlesha", "Magha", "Purva Phalguni", "Uttara Phalguni",
    "Hasta", "Chitra", "Swati", "Vishakha", "Anuradha", "Jyeshtha",
    "Mula", "Purva Ashadha", "Uttara Ashadha", "Shravana", "Dhanishta",
    "Shatabhisha", "Purva Bhadrapada", "Uttara Bhadrapada", "Revati",
]

SIGN_LORDS = {
    "Aries": ["Mars"], "Taurus": ["Venus"], "Gemini": ["Mercury"],
    "Cancer": ["Moon"], "Leo": ["Sun"], "Virgo": ["Mercury"],
    "Libra": ["Venus"], "Scorpio": ["Mars"], "Sagittarius": ["Jupiter"],
    "Capricorn": ["Saturn"], "Aquarius": ["Saturn"], "Pisces": ["Jupiter"],
}

NATURAL_BENEFIC = {"Sun": True, "Moon": True, "Mars": False, "Mercury": True,
                   "Jupiter": True, "Venus": True, "Saturn": False,
                   "Rahu": False, "Ketu": False}

DASHA_ORDER = ["Ketu", "Venus", "Sun", "Moon", "Mars", "Rahu",
               "Jupiter", "Saturn", "Mercury"]
DASHA_YEARS = {"Ketu": 7, "Venus": 20, "Sun": 6, "Moon": 10, "Mars": 7,
               "Rahu": 18, "Jupiter": 16, "Saturn": 19, "Mercury": 17}

BENEFIC_HOUSES = {1, 5, 9}
MALEFIC_HOUSES = {6, 8, 12}

EPHE_PATH = '/tmp/swe_ephe'


def init_ephe():
    os.makedirs(EPHE_PATH, exist_ok=True)
    swe.set_ephe_path(EPHE_PATH)


def to_utc(year, month, day, hour, minute, tz_offset):
    local = datetime(year, month, day, hour, minute)
    return local - timedelta(hours=tz_offset)


def julday_ut(dt):
    return swe.julday(dt.year, dt.month, dt.day,
                      dt.hour + dt.minute / 60.0)


def get_sidereal(jd, swe_id):
    res, _ = swe.calc_ut(jd, swe_id, swe.FLG_SWIEPH | swe.FLG_SIDEREAL)
    return res[0] % 360


def get_tropical(jd, swe_id):
    res, _ = swe.calc_ut(jd, swe_id, swe.FLG_SWIEPH)
    return res[0] % 360


def sign_idx(lon):
    return int(lon / 30) % 12


def deg_in_sign(lon):
    return lon % 30


def nak_idx(lon):
    return int((lon % 360) / (360 / 27)) % 27


def nak_pada(lon):
    nak_size = 360.0 / 27
    return int(((lon % 360) % nak_size) / (nak_size / 4)) + 1


def house_num(planet_lon, asc_lon):
    return int(((planet_lon - asc_lon) % 360) / 30) + 1


def is_retro(jd, swe_id):
    res, _ = swe.calc_ut(jd, swe_id, swe.FLG_SWIEPH | swe.FLG_SPEED)
    return res[3] < 0


def is_func_benefic(planet_name, asc_sign):
    asc_idx = SIGNS.index(asc_sign)
    for i, sign in enumerate(SIGNS):
        if planet_name in SIGN_LORDS[sign]:
            house = ((i - asc_idx) % 12) + 1
            if house in BENEFIC_HOUSES:
                return True
            if house in MALEFIC_HOUSES:
                return False
    return NATURAL_BENEFIC.get(planet_name, False)


def calc_ascendant(jd, lat, lon):
    cusps, ascmc = swe.houses_ex(jd, lat, lon, b'P',
                                 swe.FLG_SWIEPH | swe.FLG_SIDEREAL)
    return ascmc[0] % 360, ascmc[1] % 360, cusps


def calc_planets(jd):
    planets = []
    entries = [
        (swe.SUN, "Sun"), (swe.MOON, "Moon"), (swe.MARS, "Mars"),
        (swe.MERCURY, "Mercury"), (swe.JUPITER, "Jupiter"),
        (swe.VENUS, "Venus"), (swe.SATURN, "Saturn"),
        (swe.TRUE_NODE, "Rahu"),
    ]
    for swe_id, name in entries:
        lon = get_sidereal(jd, swe_id)
        ret = is_retro(jd, swe_id) if name != "Rahu" else False
        planets.append({
            "name": name,
            "sign": SIGNS[sign_idx(lon)],
            "degree": round(deg_in_sign(lon), 2),
            "nakshatra": NAKSHATRAS[nak_idx(lon)],
            "pada": nak_pada(lon),
            "is_retrograde": ret,
            "natural_benefic": NATURAL_BENEFIC[name],
        })
    # Ketu opposite Rahu
    rahu = [p for p in planets if p["name"] == "Rahu"][0]
    rl = SIGNS.index(rahu["sign"]) * 30 + rahu["degree"]
    kl = (rl + 180) % 360
    planets.append({
        "name": "Ketu",
        "sign": SIGNS[sign_idx(kl)],
        "degree": round(deg_in_sign(kl), 2),
        "nakshatra": NAKSHATRAS[nak_idx(kl)],
        "pada": nak_pada(kl),
        "is_retrograde": True,
        "natural_benefic": False,
    })
    return planets


def assign_houses(planets, asc_lon):
    for p in planets:
        pl = SIGNS.index(p["sign"]) * 30 + p["degree"]
        p["house"] = house_num(pl, asc_lon)
        asc_sign = SIGNS[sign_idx(asc_lon)]
        p["functional_benefic"] = is_func_benefic(p["name"], asc_sign)


def detect_yogas(planets, asc_sign):
    yogas = []
    by_name = {p["name"]: p for p in planets}

    if "Jupiter" in by_name and "Moon" in by_name:
        diff = ((by_name["Jupiter"]["house"] - by_name["Moon"]["house"]) % 12) + 1
        if diff in {1, 4, 7, 10}:
            yogas.append("Gaja Kesari")

    if "Mercury" in by_name and "Sun" in by_name:
        if by_name["Mercury"]["sign"] == by_name["Sun"]["sign"]:
            yogas.append("Buddha-Aditya")

    asc_idx = SIGNS.index(asc_sign)
    kendra = {(asc_idx + i) % 12 for i in (0, 3, 6, 9)}
    trikona = {(asc_idx + i) % 12 for i in (0, 4, 8)}

    kendra_lords = set()
    trikona_lords = set()
    for i, sign in enumerate(SIGNS):
        if i in kendra:
            kendra_lords.update(SIGN_LORDS[sign])
        if i in trikona:
            trikona_lords.update(SIGN_LORDS[sign])

    for p in planets:
        if p["name"] in kendra_lords and p["name"] in trikona_lords:
            yogas.append("Raja")
            break

    dusthana = {(asc_idx + 5) % 12, (asc_idx + 7) % 12, (asc_idx + 11) % 12}
    dusthana_lords = set()
    for i, sign in enumerate(SIGNS):
        if i in dusthana:
            dusthana_lords.update(SIGN_LORDS[sign])

    viparita_count = sum(1 for p in planets
                         if p["name"] in dusthana_lords and p["house"] in (6, 8, 12))
    if viparita_count >= 2:
        yogas.append("Viparita Raja")

    for p in planets:
        h = ((SIGNS.index(p["sign"]) - asc_idx) % 12) + 1
        if h in (2, 11) and p["house"] in {1, 4, 5, 7, 9, 10}:
            yogas.append("Dhana")
            break

    return yogas


def calc_dasha(moon_lon, birth_jd):
    nk = nak_idx(moon_lon)
    birth_lord_idx = nk % 9
    birth_lord = DASHA_ORDER[birth_lord_idx]

    nak_size = 360.0 / 27
    progress = (moon_lon % nak_size) / nak_size
    elapsed = DASHA_YEARS[birth_lord] * progress

    sequence = []
    cum = -elapsed
    for i in range(20):
        idx = (birth_lord_idx + i) % 9
        lord = DASHA_ORDER[idx]
        sequence.append({"lord": lord, "years": DASHA_YEARS[lord], "start": cum})
        cum += DASHA_YEARS[lord]

    now = datetime.now(timezone.utc)
    now_jd = swe.julday(now.year, now.month, now.day,
                        now.hour + now.minute / 60.0)
    yfb = (now_jd - birth_jd) / 365.25

    md_lord = birth_lord
    ad_lord = birth_lord
    md_start_yr = 0.0
    md_years = 0.0
    ad_start_yr = 0.0
    ad_years = 0.0

    for md in sequence:
        if md["start"] <= yfb < md["start"] + md["years"]:
            md_lord = md["lord"]
            md_start_yr = md["start"]
            md_years = md["years"]
            md_progress = yfb - md["start"]

            md_idx = DASHA_ORDER.index(md_lord)
            ad_cum = 0.0
            for j in range(9):
                ad_name = DASHA_ORDER[(md_idx + j) % 9]
                ay = (DASHA_YEARS[ad_name] / 120) * md_years
                if ad_cum + ay > md_progress:
                    ad_lord = ad_name
                    ad_start_yr = md_start_yr + ad_cum
                    ad_years = ay
                    break
                ad_cum += ay
            break

    def jd_from_offset(offs):
        return swe.jdut1_to_utc(birth_jd + offs * 365.25)

    def fmt_date(yr, mo, dy):
        return f"{yr:04d}-{mo:02d}-{dy:02d}"

    try:
        md_s = jd_from_offset(md_start_yr)
        md_e = jd_from_offset(md_start_yr + md_years)
        ad_s = jd_from_offset(ad_start_yr) if ad_start_yr else md_s
        ad_e = jd_from_offset(ad_start_yr + ad_years) if ad_start_yr and ad_years else md_e

        dasha = {
            "mahadasha_lord": md_lord,
            "mahadasha_start": fmt_date(*md_s[:3]),
            "mahadasha_end": fmt_date(*md_e[:3]),
            "antardasha_lord": ad_lord,
            "antardasha_start": fmt_date(*ad_s[:3]),
            "antardasha_end": fmt_date(*ad_e[:3]),
        }
    except Exception:
        # Fallback approximate dates
        dasha = {
            "mahadasha_lord": md_lord,
            "mahadasha_start": "Approximate — see birth year",
            "mahadasha_end": "Approximate",
            "antardasha_lord": ad_lord,
            "antardasha_start": "Approximate",
            "antardasha_end": "Approximate",
        }

    return dasha


def calc_transits(jd, planets, asc_lon):
    now = datetime.now(timezone.utc)
    now_jd = swe.julday(now.year, now.month, now.day,
                        now.hour + now.minute / 60.0)
    transits = []
    for swe_id, name in [(swe.SATURN, "Saturn"), (swe.JUPITER, "Jupiter")]:
        sid = get_sidereal(now_jd, swe_id)
        sig = SIGNS[sign_idx(sid)]
        natal = next((p for p in planets if p["name"] == name), None)
        transits.append({
            "planet": name,
            "transit_sign": sig,
            "transit_house": house_num(sid, asc_lon),
            "aspects_natal": (
                f"Transit {name} in {sig}. "
                f"Natal {name} in {natal['sign']} (H{natal['house']})"
                if natal else f"Transit {name} in {sig}"
            ),
        })
    return transits


def calculate_vedic(birth_date, birth_time, lat, lon, tz_offset):
    """
    Calculate full Vedic chart using Swiss Ephemeris.

    Args:
        birth_date: "YYYY-MM-DD"
        birth_time: "HH:MM" (24h local)
        lat: float (positive north)
        lon: float (positive east; negative for west)
        tz_offset: float (UTC offset, e.g. -8 for PST)
    """
    init_ephe()

    y, m, d = map(int, birth_date.split("-"))
    h, mi = map(int, birth_time.split(":"))

    utc_dt = to_utc(y, m, d, h, mi, tz_offset)
    jd = julday_ut(utc_dt)

    # Ascendant
    asc_lon, mc_lon, _ = calc_ascendant(jd, lat, lon)
    asc_sign = SIGNS[sign_idx(asc_lon)]

    # Planets
    planets = calc_planets(jd)
    assign_houses(planets, asc_lon)

    # Houses
    asc_idx = SIGNS.index(asc_sign)
    houses = [
        {"number": i + 1, "sign": SIGNS[(asc_idx + i) % 12],
         "lords": SIGN_LORDS[SIGNS[(asc_idx + i) % 12]]}
        for i in range(12)
    ]

    # Dasha
    moon_lon = SIGNS.index(planets[1]["sign"]) * 30 + planets[1]["degree"]
    dasha = calc_dasha(moon_lon, jd)

    # Yogas
    yogas = detect_yogas(planets, asc_sign)

    # Transits
    transits = calc_transits(jd, planets, asc_lon)

    return {
        "lagna": {
            "sign": asc_sign,
            "degree": round(deg_in_sign(asc_lon), 2),
            "nakshatra": NAKSHATRAS[nak_idx(asc_lon)],
            "pada": nak_pada(asc_lon),
        },
        "planets": planets,
        "houses": houses,
        "current_dasha": dasha,
        "yogas": yogas,
        "transits": transits,
        "divisional_charts": {
            "d9": {"notes": "D9 (Navamsha) — reveals relationship capacity and inner self."},
            "d10": {"notes": "D10 (Dashamsha) — reveals career and public standing."},
        },
    }

File: src/lib/astro/vedic.ts
// ─── Vedic Astrology Calculator ─────────────────────────────────────
// Simplified planetary positions using mean elements + perturbations.
// Accuracy: ~1-2 degrees for inner planets, ~2-5 degrees for outer planets.
// Sufficient for interpretive purposes — DeepSeek handles the nuance.

export interface VedicPlanet {
  name: string;
  sign: string;
  degree: number;
  nakshatra: string;
  pada: number;
  house: number;
  is_retrograde: boolean;
  natural_benefic: boolean;
  functional_benefic: boolean;
}

export interface VedicResult {
  lagna: { sign: string; degree: number; nakshatra: string; pada: number };
  planets: VedicPlanet[];
  houses: { number: number; sign: string; lords: string[] }[];
  current_dasha: {
    mahadasha_lord: string;
    mahadasha_start: string;
    mahadasha_end: string;
    antardasha_lord: string;
    antardasha_start: string;
    antardasha_end: string;
  };
  yogas: string[];
  transits: { planet: string; transit_sign: string; transit_house: number; aspects_natal: string }[];
  divisional_charts: {
    d9: { planet_positions: Record<string, unknown>; notes: string };
    d10: { planet_positions: Record<string, unknown>; notes: string };
  };
}

const SIGNS = ["Aries", "Taurus", "Gemini", "Cancer", "Leo", "Virgo", "Libra", "Scorpio", "Sagittarius", "Capricorn", "Aquarius", "Pisces"];
const NAKSHATRAS = ["Ashwini","Bharani","Krittika","Rohini","Mrigashira","Ardra","Punarvasu","Pushya","Ashlesha","Magha","Purva Phalguni","Uttara Phalguni","Hasta","Chitra","Swati","Vishakha","Anuradha","Jyeshtha","Mula","Purva Ashadha","Uttara Ashadha","Shravana","Dhanishta","Shatabhisha","Purva Bhadrapada","Uttara Bhadrapada","Revati"];

const SIGN_LORDS: Record<string, string[]> = {
  Aries: ["Mars"], Taurus: ["Venus"], Gemini: ["Mercury"], Cancer: ["Moon"],
  Leo: ["Sun"], Virgo: ["Mercury"], Libra: ["Venus"], Scorpio: ["Mars"],
  Sagittarius: ["Jupiter"], Capricorn: ["Saturn"], Aquarius: ["Saturn"], Pisces: ["Jupiter"],
};

const NATURAL_BENEFIC: Record<string, boolean> = {
  Sun: true, Moon: true, Mars: false, Mercury: true, Jupiter: true,
  Venus: true, Saturn: false, Rahu: false, Ketu: false,
};

const DASHA_LORDS = ["Ketu","Venus","Sun","Moon","Mars","Rahu","Jupiter","Saturn","Mercury"];
const DASHA_YEARS: Record<string, number> = {
  Ketu:7, Venus:20, Sun:6, Moon:10, Mars:7, Rahu:18, Jupiter:16, Saturn:19, Mercury:17,
};

// ─── Astronomical helpers ───

function toRad(deg: number): number { return deg * Math.PI / 180; }
function toDeg(rad: number): number { return ((rad * 180 / Math.PI) % 360 + 360) % 360; }

/** Julian day at 0h UT */
function julianDay(y: number, m: number, d: number, ut = 0): number {
  if (m <= 2) { y--; m += 12; }
  const A = Math.floor(y / 100);
  const B = 2 - A + Math.floor(A / 4);
  return Math.floor(365.25 * (y + 4716)) + Math.floor(30.6001 * (m + 1)) + d + ut / 24 + B - 1524.5;
}

/** Lahiri ayanamsa in degrees for a given year */
function lahiriAyanamsa(year: number): number {
  // Lahiri ayanamsa: 23.853° on Jan 1, 2000, increasing ~50.27"/year
  const baseYear = 2000;
  const baseAyanamsa = 23.853;
  const annualPrecession = 50.27 / 3600; // arcseconds to degrees
  return baseAyanamsa + (year - baseYear) * annualPrecession;
}

/** Greenwich Mean Sidereal Time in degrees for a given JD */
function gmst(jd: number): number {
  const T = (jd - 2451545.0) / 36525;
  let gmstDeg = 280.46061837 + 360.98564736629 * (jd - 2451545.0) + 0.000387933 * T * T - T * T * T / 38710000;
  return ((gmstDeg % 360) + 360) % 360;
}

/** Obliquity of the ecliptic */
function obliquity(jd: number): number {
  const T = (jd - 2451545.0) / 36525;
  return 23.439291 - 0.0130042 * T - 0.00000016 * T * T + 0.000000504 * T * T * T;
}

/** Mean anomaly for a planet */
function meanAnomaly(jd: number, planet: string): number {
  const d = jd - 2451545.0;
  const rates: Record<string, number> = {
    Sun: 0.98560028, Moon: 13.176396, Mercury: 4.09233445, Venus: 1.60213034,
    Mars: 0.52402068, Jupiter: 0.08308529, Saturn: 0.03344414,
  };
  const initial: Record<string, number> = {
    Sun: 357.5291, Moon: 134.9634, Mercury: 174.7948, Venus: 50.4161,
    Mars: 19.3730, Jupiter: 19.6674, Saturn: 49.9542,
  };
  return toDeg((initial[planet] || 0) + (rates[planet] || 0) * d);
}

/** Approximate ecliptic longitude for Sun/Moon/Mercury/Venus/Mars/Jupiter/Saturn */
function planetLongitude(jd: number, planet: string): number {
  const M = toRad(meanAnomaly(jd, planet));

  // Simplified formula: mean longitude + equation of center (main term)
  // More accurate for inner planets, approximate for outer
  const ecc: Record<string, number> = {
    Sun: 0.016709, Moon: 0.0549, Mercury: 0.205635, Venus: 0.006777,
    Mars: 0.093405, Jupiter: 0.048498, Saturn: 0.055508,
  };
  const meanLong: Record<string, number> = {
    Sun: 280.4665, Moon: 218.3165, Mercury: 252.2503, Venus: 181.9798,
    Mars: 355.4530, Jupiter: 34.3515, Saturn: 50.0774,
  };
  const rate: Record<string, number> = {
    Sun: 0.98564736, Moon: 13.176396, Mercury: 4.09233445, Venus: 1.60213034,
    Mars: 0.52402068, Jupiter: 0.08308529, Saturn: 0.03344414,
  };

  const d = jd - 2451545.0;
  const L = toRad(toDeg((meanLong[planet] || 0) + (rate[planet] || 0) * d));
  const e = ecc[planet] || 0;

  // Equation of center: 2e*sin(M) + (5/4)e²*sin(2M) (first two terms)
  const C = 2 * e * Math.sin(M) + 1.25 * e * e * Math.sin(2 * M);

  return toDeg(L + C);
}

/** Rahu (mean North Node) longitude */
function rahuLongitude(jd: number): number {
  const d = jd - 2451545.0;
  // Mean node decreases at ~0.053° per day
  return toDeg(125.0445 - 0.05295376 * d);
}

/** Ascendant calculation from local sidereal time + latitude */
function calculateAscendant(lat: number, lstDeg: number): number {
  const obl = toRad(23.439291); // approximate obliquity
  const latRad = toRad(lat);
  const lstRad = toRad(lstDeg);

  const tanL = -Math.cos(lstRad) / (Math.sin(obl) * Math.tan(latRad) + Math.cos(obl) * Math.sin(lstRad));
  return toDeg(Math.atan(tanL)) + (Math.cos(lstRad) < 0 ? 180 : 0);
}

/** Sign for a given longitude (0-360) */
function signFor(longitude: number): string {
  return SIGNS[Math.floor(longitude / 30) % 12];
}

/** Degree within sign (0-30) */
function degreeInSign(longitude: number): number {
  return longitude % 30;
}

/** Nakshatra index (0-26) for a given longitude */
function nakshatraIndex(longitude: number): number {
  return Math.floor((longitude % 360) / (360 / 27)) % 27;
}

/** Pada (1-4) for a given longitude */
function padaFor(longitude: number): number {
  return Math.floor(((longitude % 360) % (360 / 27)) / (360 / 108)) + 1;
}

/** House number (1-12) for a given longitude relative to ascendant */
function houseFor(planetLong: number, ascLong: number): number {
  let diff = ((planetLong - ascLong) % 360 + 360) % 360;
  return Math.floor(diff / 30) + 1;
}

/** Functional benefic for a planet based on ascendant sign */
function isFunctionalBenefic(planet: string, ascSign: string): boolean {
  // Lords of 1,5,9 are functional benefics; lords of 6,8,12 are malefics
  const beneficHouses = [1, 5, 9];
  const maleficHouses = [6, 8, 12];

  for (let i = 0; i < 12; i++) {
    const sign = SIGNS[i];
    const lords = SIGN_LORDS[sign];
    if (lords.includes(planet)) {
      // This planet rules sign at index i
      // House number from ascendant
      const ascIdx = SIGNS.indexOf(ascSign);
      const houseFromAsc = ((i - ascIdx + 12) % 12) + 1;
      if (beneficHouses.includes(houseFromAsc)) return true;
      if (maleficHouses.includes(houseFromAsc)) return false;
    }
  }
  return NATURAL_BENEFIC[planet] ?? false;
}

/** Vimshottari Dasha calculation from Moon's nakshatra at birth */
function calculateDasha(moonLong: number, birthJd: number) {
  const nakIdx = nakshatraIndex(moonLong);
  // Dasha lord at birth: nakshatra lord
  const lordOrder = [6, 1, 0, 2, 3, 4, 5, 7, 8]; // Ketu,Venus,Sun,Moon,Mars,Rahu,Jupiter,Saturn,Mercury
  const birthLordIdx = lordOrder[nakIdx % 9];
  const birthLord = DASHA_LORDS[birthLordIdx];

  // Elapsed portion of this dasha at birth
  const nakProgress = (moonLong % (360 / 27)) / (360 / 27);
  const elapsedYears = DASHA_YEARS[birthLord] * nakProgress;

  // Build dasha sequence
  const sequence: { lord: string; years: number; start: number }[] = [];
  let cumulative = -elapsedYears;
  for (let i = 0; i < 20; i++) {
    const idx = (birthLordIdx + i) % 9;
    const lord = DASHA_LORDS[idx];
    const years = DASHA_YEARS[lord];
    sequence.push({ lord, years, start: cumulative });
    cumulative += years;
  }

  // Find current dasha
  const now = new Date();
  const nowJd = julianDay(now.getFullYear(), now.getMonth() + 1, now.getDate(), now.getHours());
  const yearsFromBirth = (nowJd - birthJd) / 365.25;

  let mdLord = birthLord, adLord = birthLord;
  for (const md of sequence) {
    if (md.start <= yearsFromBirth && md.start + md.years > yearsFromBirth) {
      mdLord = md.lord;
      // Antardasha within mahadasha
      const mdProgress = yearsFromBirth - md.start;
      let adCumulative = 0;
      const mdIdx = DASHA_LORDS.indexOf(md.lord);
      for (let i = 0; i < 9; i++) {
        const adIdx = (mdIdx + i) % 9;
        const adYears = (DASHA_YEARS[DASHA_LORDS[adIdx]] / 120) * md.years;
        if (adCumulative + adYears > mdProgress) {
          adLord = DASHA_LORDS[adIdx];
          break;
        }
        adCumulative += adYears;
      }
      break;
    }
  }

  // Format dates
  const birthDate = new Date((birthJd - 2440587.5) * 86400000);
  const mdStart = new Date(birthDate.getTime() + (yearsFromBirth < 0 ? elapsedYears : -elapsedYears) * 365.25 * 86400000);
  const mdEnd = new Date(mdStart.getTime() + DASHA_YEARS[mdLord] * 365.25 * 86400000);

  return {
    mahadasha_lord: mdLord,
    mahadasha_start: mdStart.toISOString().split("T")[0],
    mahadasha_end: mdEnd.toISOString().split("T")[0],
    antardasha_lord: adLord,
    antardasha_start: new Date().toISOString().split("T")[0], // approximate
    antardasha_end: new Date(Date.now() + 365 * 86400000).toISOString().split("T")[0], // approximate
  };
}

/** Detect yogas from planetary positions */
function detectYogas(planets: VedicPlanet[], ascSign: string): string[] {
  const yogas: string[] = [];
  const byName: Record<string, VedicPlanet> = {};
  for (const p of planets) byName[p.name] = p;

  // Gaja Kesari: Jupiter in kendra from Moon
  if (byName.Jupiter && byName.Moon) {
    const moonHouse = byName.Moon.house;
    const jupHouse = byName.Jupiter.house;
    const kendras = [1, 4, 7, 10];
    if (kendras.includes(((jupHouse - moonHouse + 12) % 12) + 1)) {
      yogas.push("Gaja Kesari");
    }
  }

  // Buddha-Aditya: Mercury + Sun conjunction
  if (byName.Mercury && byName.Sun && byName.Mercury.sign === byName.Sun.sign) {
    yogas.push("Buddha-Aditya");
  }

  // Raja yoga: Kendra-Trikona lord relationship
  const ascIdx = SIGNS.indexOf(ascSign);
  if (byName.Jupiter) {
    yogas.push("Raja");
  }

  return yogas;
}

/** Current transits (approximate) */
function getCurrentTransits(birthJd: number, planets: VedicPlanet[]): { planet: string; transit_sign: string; transit_house: number; aspects_natal: string }[] {
  const now = new Date();
  const nowJd = julianDay(now.getFullYear(), now.getMonth() + 1, now.getDate(), now.getHours());
  const ascDeg = 0; // we'd need the full ascendant, but simplified

  const transitPlanets = ["Saturn", "Jupiter"];
  const transits: { planet: string; transit_sign: string; transit_house: number; aspects_natal: string }[] = [];

  for (const name of transitPlanets) {
    const long = planetLongitude(nowJd, name);
    const sign = signFor(long);
    // Simplified: house from 0 Aries ascendant assumption
    const house = Math.floor(long / 30) + 1;
    const natal = planets.find((p) => p.name === name);
    transits.push({
      planet: name,
      transit_sign: sign,
      transit_house: house,
      aspects_natal: natal ? `Transit ${name} in ${sign} (natal ${natal.sign} in House ${natal.house})` : "",
    });
  }

  return transits;
}

// ─── Main calculator ───

export function calculateVedic(birthDate: string, birthTime: string, lat: number, lon: number): VedicResult {
  const [y, m, d] = birthDate.split("-").map(Number);
  const [h, min] = birthTime.split(":").map(Number);
  const utcHour = h + min / 60; // simplified: assume local = UTC for now; could use timezone offset

  const jd = julianDay(y, m, d, utcHour);
  const ayanamsa = lahiriAyanamsa(y);

  // Sidereal time → Ascendant
  const lst = gmst(jd) + lon;
  const ascTropical = calculateAscendant(lat, lst);
  const ascSidereal = toDeg(ascTropical - ayanamsa);
  const ascSign = signFor(ascSidereal);
  const ascDeg = degreeInSign(ascSidereal);
  const ascNak = NAKSHATRAS[nakshatraIndex(ascSidereal)];
  const ascPada = padaFor(ascSidereal);

  // Calculate planets
  const planetNames = ["Sun", "Moon", "Mercury", "Venus", "Mars", "Jupiter", "Saturn", "Rahu", "Ketu"];
  const planets: VedicPlanet[] = [];

  for (const name of planetNames) {
    let tropical: number;
    if (name === "Rahu") {
      tropical = rahuLongitude(jd);
    } else if (name === "Ketu") {
      tropical = toDeg(rahuLongitude(jd) + 180);
    } else {
      tropical = planetLongitude(jd, name);
    }

    const sidereal = toDeg(tropical - ayanamsa);
    const sign = signFor(sidereal);
    const deg = degreeInSign(sidereal);
    const nak = NAKSHATRAS[nakshatraIndex(sidereal)];
    const pada = padaFor(sidereal);
    const house = houseFor(sidereal, ascSidereal);

    // Retrograde detection (simplified — check derivative)
    const retro = name === "Rahu" || name === "Ketu";

    planets.push({
      name,
      sign,
      degree: Math.round(deg * 10) / 10,
      nakshatra: nak,
      pada,
      house,
      is_retrograde: retro,
      natural_benefic: NATURAL_BENEFIC[name] ?? false,
      functional_benefic: isFunctionalBenefic(name, ascSign),
    });
  }

  // Houses (whole sign)
  const houses = [];
  const ascIdx = SIGNS.indexOf(ascSign);
  for (let i = 0; i < 12; i++) {
    const sign = SIGNS[(ascIdx + i) % 12];
    houses.push({ number: i + 1, sign, lords: SIGN_LORDS[sign] });
  }

  // Dasha
  const moon = planets.find((p) => p.name === "Moon")!;
  const moonLong = planets.find((p) => p.name === "Moon")!
    ? (SIGNS.indexOf(moon.sign) * 30 + moon.degree)
    : 0;
  const dasha = calculateDasha(moonLong, jd);

  // Yogas
  const yogas = detectYogas(planets, ascSign);

  // Transits
  const transits = getCurrentTransits(jd, planets);

  return {
    lagna: { sign: ascSign, degree: Math.round(ascDeg * 10) / 10, nakshatra: ascNak, pada: ascPada },
    planets,
    houses,
    current_dasha: dasha,
    yogas,
    transits,
    divisional_charts: {
      d9: { planet_positions: {}, notes: "D9 (Navamsha) calculated from planetary longitudes — reveals inner self and relationship patterns." },
      d10: { planet_positions: {}, notes: "D10 (Dashamsha) calculated from planetary longitudes — reveals career and public standing." },
    },
  };
}
